sql-non-prepared-plan-cache.md

```diff
@@ -0,0 +1,123 @@
+---
+title: 非 Prepared 语句执行计划缓存
+aliases: ['/docs-cn/dev/sql-non-prepare-plan-cache/','zh/tidb/dev/sql-non-prepare-plan-cache']
+---
+
+# 非 Prepared 语句执行计划缓存
+
+> **警告：**
+>
+> 该功能为实验特性，不推荐在生产环境中使用。
+
+对于一部分非 Prepared 语句，TiDB 也能像 [`Prepare` / `Execute` 语句](sql-prepared-plan-cache.md) 一样支持执行计划缓存，可以让这些语句跳过优化器阶段，以获得性能上的提升。
+
+此功能基本原理如下：
+
+1. 打开此功能后，TiDB 首先根据 AST 对查询进行参数化，如将 `select * from t where b<10 and a=1` 参数化为 `select * from t where b<? and a=?`；
+2. 然后用参数化后的查询，在 Non-Prepared Plan Cache 中进行查找；
+3. 如果能找到可以直接复用的计划，则直接使用，并跳过整个优化过程；
+4. 如果找不到可以直接复用的计划，则继续进行查询优化，并在最后把生成的计划放回到 Cache 中，以便下次进行复用；
+
+Non-Prepared Plan Cache 为 Session 级别，且和 [Prepared Plan Cache](sql-prepared-plan-cache.md) 相互独立，其中缓存的 Plan 互不影响。
+
+目前可以通过 `tidb_enable_non_prepared_plan_cache` 来打开和关闭此项功能，同时通过 `tidb_non_prepared_plan_cache_size` 来控制 Non-Prepared Plan Cache 的大小，当缓存的计划数超过 `tidb_non_prepared_plan_cache_size` 时，会使用 LRU 策略来进行逐出。
+
+## 实例
+
+下面是一些实例：
+
+{{< copyable "sql" >}}
+
+```sql
+mysql> create table t (a int, b int, key(b));           -- 创建用于测试的 table
+Query OK, 0 rows affected (0.06 sec)
+
+mysql> set tidb_enable_non_prepared_plan_cache=true;    -- 打开 Non-Prepared Plan Cache
+Query OK, 0 rows affected (0.00 sec)
+
+mysql> select * from t where b < 10 and a=1;            -- 第一次执行查询
+Empty set (0.00 sec)
+
+mysql> select * from t where b < 5 and a=2;             -- 第二次执行查询
+Empty set (0.00 sec)
+
+mysql> select @@last_plan_from_cache;                   -- 可以看到第二次执行的执行计划来自于 Cache
++------------------------+
+| @@last_plan_from_cache |
++------------------------+
+|                      1 |
++------------------------+
+1 row in set (0.00 sec)
+```
```

Comment:
````suggestion
1. 创建用于测试的表 `t`：

    ```sql
    CREATE TABLE t (a INT, b INT, KEY(b));
    ```

2. 开启 Non-Prepared Plan Cache：

    ```sql
    SET tidb_enable_non_prepared_plan_cache = true;
    ```

3. 依次执行以下查询：

    ```sql
    SELECT * FROM t WHERE b < 10 AND a = 1;
    SELECT * FROM t WHERE b < 5 AND a = 2;
    ```

4. 查看第二个查询语句是否命中 cache：

    ```sql
    SELECT @@last_plan_from_cache;
    ```

    输出结果中 `last_plan_from_cache` 的值为 `1`，表示第二次执行的查询计划来自于 cache。

    ```sql
    +------------------------+
    | @@last_plan_from_cache |
    +------------------------+
    |                      1 |
    +------------------------+
    1 row in set (0.00 sec)
    ```
````